design.md

```diff
@@ -0,0 +1,109 @@
+Many of these instructors do not have traditional training in computer science or statistics, and are learning data science skills on the fly, as they develop and teach their courses.
+
+Jupyter notebooks are widely employed in classes at Cal because they provide a natural environment for introducing data science skills to students.  Critically, they enable a browser-based interface to computation in the cloud, meaning that students only need a browser to start programming and interacting with data.  
+
+However, the tools currently available for instructors require considerably more experience with computational and programming tools. In order to include and support this multidisciplinary group of instructors who are developing a rich ensemble of classes, particularly from disciplines currently underrepresented in data science, the tools available must be designed to facilitate the work of instructors whose research and educational backgrounds do not include, for instance, tools for version control. We focus here on a particular problem that they encounter early in developing and teaching the courses: managing the notebooks, data, and other files they develop and distribute to students.
+
+For the instructors' current workflow, see https://github.com/data-8/connector-instructors#workflow 
+
+# Summary of Solution #
+
+Implement Dropbox-like synchronization of specific folders, accessible via the Jupyter Notebook interface. We will initially provide a GitHub backed implementation, but in the future other backends could be made available (e.g. Dropbox, generic git). We will explicitly avoid making a true Git GUI client - the user's mental model should be 'this is similar to Dropbox.'
+
+# Implementation details #
+
+This would be implemented as a Jupyter notebook backend extension. We'd use a very small subset of git commands, tied specifically into GitHub, to provide this synchronizing functionality.
+
+## Synchronization ##
+
+The actual synchronization would be handled by the python part of the nbextension, via logic similar to https://gist.github.com/yuvipanda/6cdeab86ccaa15687b25. 
+
+1. Every 30s (or other configurable interval)...
+2. We perform an automatic commit of everything in the Synchronized Folder (with a not very useful commit message)
+3. Do a git pull with an automerge that's guaranteed to never conflict, by simply opinionatedly choosing one side of the conflict (by default, the most recent version based on timestamp)
+4. Push to master
```

Comment:
According to this as it currently stands, all users of a Jupyterhub system will get all the sync'ed folders. Is that right?

That sounds fine to me as long as we allow for more granular control later (eg. instructor A distributes folders to students 1 and 2, instructor B distributes folders to students 2 and 3).